applications/ShapeOptimizationApplication/custom_utilities/mapping/mapper_vertex_morphing_symmetric.cpp

```diff
@@ -0,0 +1,363 @@
+// ==============================================================================
+//  KratosShapeOptimizationApplication
+//
+//  License:         BSD License
+//                   license: ShapeOptimizationApplication/license.txt
+//
+//  Main authors:    Armin Geiser, https://github.com/armingeiser
+//
+// ==============================================================================
+
+// ------------------------------------------------------------------------------
+// System includes
+// ------------------------------------------------------------------------------
+#include <iostream>
+#include <string>
+#include <algorithm>
+
+// ------------------------------------------------------------------------------
+// Project includes
+// ------------------------------------------------------------------------------
+#include "includes/define.h"
+#include "includes/model_part.h"
+#include "spatial_containers/spatial_containers.h"
+#include "utilities/builtin_timer.h"
+#include "spaces/ublas_space.h"
+#include "mapper_base.h"
+#include "filter_function.h"
+#include "symmetry_base.h"
+#include "symmetry_plane.h"
+#include "symmetry_revolution.h"
+#include "mapper_vertex_morphing_symmetric.h"
+
+// ==============================================================================
+
+namespace Kratos
+{
+
+void MapperVertexMorphingSymmetric::Initialize()
+{
+    BuiltinTimer timer;
+    KRATOS_INFO("ShapeOpt") << "Starting initialization of mapper..." << std::endl;
+
+    CreateFilterFunction();
+
+    mIsMappingInitialized = true;
+
+    Update();
+
+    KRATOS_INFO("ShapeOpt") << "Finished initialization of mapper in " << timer.ElapsedSeconds() << " s." << std::endl;
+}
+
+void MapperVertexMorphingSymmetric::Map( const Variable<array_3d> &rOriginVariable, const Variable<array_3d> &rDestinationVariable)
+{
+    if (mIsMappingInitialized == false)
+        Initialize();
+
+    BuiltinTimer timer;
+    KRATOS_INFO("") << std::endl;
+    KRATOS_INFO("ShapeOpt") << "Starting mapping of " << rOriginVariable.Name() << "..." << std::endl;
+
+    Vector values_origin(mrOriginModelPart.Nodes().size()*3);
+    Vector values_destination(mrDestinationModelPart.Nodes().size()*3);
+
+    // Prepare vectors for mapping
+    values_origin.clear();
+    values_destination.clear();
+
+    block_for_each(mrOriginModelPart.Nodes(), [&](const ModelPart::NodeType& rNode) {
+        const int i = rNode.GetValue(MAPPING_ID);
+        const array_3d& r_nodal_variable = rNode.FastGetSolutionStepValue(rOriginVariable);
+        values_origin[i*3+0] = r_nodal_variable[0];
+        values_origin[i*3+1] = r_nodal_variable[1];
+        values_origin[i*3+2] = r_nodal_variable[2];
+    });
+
+    // Perform mapping
+    noalias(values_destination) = prod(mMappingMatrix,values_origin);
```

Comment:
why using `prod` here and using `SparseSpaceType::TransposeMult` in the `InversMap` function ?

Reply:
ah yes I also wanted to comment this
I would use `Mult` here, it is optimized